File: sdks/java/src/main/java/ai/beater/sdk/SemConv.java
package ai.beater.sdk;

import java.util.Set;

/**
 * Beater semantic conventions: the single source of truth for span-kind and
 * attribute keys in the Java SDK.
 *
 * <p>These strings mirror the Rust normalizer in {@code crates/beater-otlp} and
 * the canonical kinds in {@code crates/beater-schema}, and stay in lockstep with
 * the Python and TypeScript SDKs. They are part of the cross-language ingest
 * contract; keep them aligned with the server.
 */
public final class SemConv {

    private SemConv() {}

    // --- OpenInference span kinds (openinference.span.kind values) ---
    public static final String AGENT_RUN = "agent.run";
    public static final String AGENT_TURN = "agent.turn";
    public static final String AGENT_PLAN = "agent.plan";
    public static final String AGENT_STEP = "agent.step";
    public static final String LLM_CALL = "llm.call";
    public static final String TOOL_CALL = "tool.call";
    public static final String MCP_REQUEST = "mcp.request";
    public static final String RETRIEVAL_QUERY = "retrieval.query";
    public static final String MEMORY_READ = "memory.read";
    public static final String MEMORY_WRITE = "memory.write";
    public static final String GUARDRAIL_CHECK = "guardrail.check";

    /** Every accepted span kind, for validation. */
    public static final Set<String> SPAN_KINDS = Set.of(
            AGENT_RUN,
            AGENT_TURN,
            AGENT_PLAN,
            AGENT_STEP,
            LLM_CALL,
            TOOL_CALL,
            MCP_REQUEST,
            RETRIEVAL_QUERY,
            MEMORY_READ,
            MEMORY_WRITE,
            GUARDRAIL_CHECK);

    // --- Canonical span attribute keys ---
    public static final String SPAN_KIND = "openinference.span.kind";
    public static final String SEQ = "beater.seq";
    public static final String RELEASE_ID = "agent.release_id";

    public static final String INPUT_VALUE = "input.value";
    public static final String OUTPUT_VALUE = "output.value";

    public static final String LLM_PROVIDER = "llm.provider";
    public static final String LLM_MODEL_NAME = "llm.model_name";
    public static final String LLM_TOKEN_PROMPT = "llm.token_count.prompt";
    public static final String LLM_TOKEN_COMPLETION = "llm.token_count.completion";
    public static final String LLM_TOKEN_REASONING = "llm.token_count.reasoning";
    public static final String LLM_TOKEN_CACHE_READ = "llm.token_count.cache_read";
    public static final String LLM_COST_MICROS = "llm.cost.amount_micros";
    public static final String LLM_COST_CURRENCY = "llm.cost.currency";

    // --- OTLP ingest headers (used to scope traces over gRPC) ---
    public static final String HEADER_TENANT = "x-beater-tenant-id";
    public static final String HEADER_PROJECT = "x-beater-project-id";
    public static final String HEADER_ENVIRONMENT = "x-beater-environment-id";
}

File: sdks/python/beater/semconv.py
"""Beater semantic conventions.

This is the SINGLE place the Python SDK defines span-kind and attribute keys.
Every wrapper (`observe`, provider wrappers, and framework callbacks) imports
from here so the instrumentation can never drift within the SDK.

These strings mirror the Rust normalizer in ``crates/beater-otlp`` and the
canonical kinds in ``crates/beater-schema``. They are part of the cross-language
ingest contract; keep them in lockstep with the server. (A future codegen step
can emit this file from the Rust source -- see ``sdks/README.md``.)
"""

from __future__ import annotations


class SpanKind:
    """OpenInference ``openinference.span.kind`` values accepted by the normalizer."""

    AGENT_RUN = "agent.run"
    AGENT_TURN = "agent.turn"
    AGENT_PLAN = "agent.plan"
    AGENT_STEP = "agent.step"
    LLM_CALL = "llm.call"
    TOOL_CALL = "tool.call"
    MCP_REQUEST = "mcp.request"
    RETRIEVAL_QUERY = "retrieval.query"
    MEMORY_READ = "memory.read"
    MEMORY_WRITE = "memory.write"
    GUARDRAIL_CHECK = "guardrail.check"


#: Every accepted span kind, for validation.
SPAN_KINDS = frozenset(
    value
    for name, value in vars(SpanKind).items()
    if not name.startswith("_") and isinstance(value, str)
)


class Attr:
    """Canonical span attribute keys."""

    SPAN_KIND = "openinference.span.kind"
    SEQ = "beater.seq"
    RELEASE_ID = "agent.release_id"

    INPUT_VALUE = "input.value"
    OUTPUT_VALUE = "output.value"

    LLM_PROVIDER = "llm.provider"
    LLM_MODEL_NAME = "llm.model_name"
    LLM_TOKEN_PROMPT = "llm.token_count.prompt"
    LLM_TOKEN_COMPLETION = "llm.token_count.completion"
    LLM_TOKEN_REASONING = "llm.token_count.reasoning"
    LLM_TOKEN_CACHE_READ = "llm.token_count.cache_read"
    LLM_COST_MICROS = "llm.cost.amount_micros"
    LLM_COST_CURRENCY = "llm.cost.currency"


#: OTLP ingest headers used to scope traces when exporting over gRPC.
HEADER_TENANT = "x-beater-tenant-id"
HEADER_PROJECT = "x-beater-project-id"
HEADER_ENVIRONMENT = "x-beater-environment-id"
